### PR TITLE
[Feature #19406] Allow declarative definition of references for rb_typed_data_struct

### PR DESCRIPTION
**Pull request by eightbitraptor**
When using rb_data_type_struct to wrap a C struct, that C struct can
contain VALUE references to other Ruby objects.

If this is the case then one must also define dmark and optionally
dcompact callbacks in order to allow these objects to be correctly
handled by the GC. This is suboptimal as it requires GC related logic to
be implemented by extension developers. This can be a cause of subtle
bugs when references are not marked of updated correctly inside these
callbacks.

This commit provides an alternative approach, useful in the simple case
where the C struct contains VALUE members (ie. there isn't any
conditional logic, or data structure manipulation required to traverse
these references).

In this case references can be defined using a declarative syntax
as a list of edges (or, pointers to references).

A flag can be set on the rb_data_type_struct to notify the GC that
declarative references are being used, and a list of those references
can be assigned to the arbitrary data pointer on the
rb_data_type_struct.

Macros are also provided for simple declaration of the reference list,
and building edges.

To avoid having to also find space in the struct to define a length for
the references list, I've chosed to always terminate the references list
with RUBY_REF_END - defined as UINTPTR_MAX. My assumption is that no
single struct will ever be large enough that UINTPTR_MAX is actually a
valid reference.

This PR also provides example implementations for `dir_data` and 
`enumerator`

## TODO (Post merge)

- [ ] Implement for more types

### CODE REVIEW

File: gc.c [hunk 1]
```diff
@@ -7208,6 +7208,47 @@ gc_mark_imemo(rb_objspace_t *objspace, VALUE obj)
...
+gc_mark_from_offset(rb_objspace_t *objspace, VALUE *data_struct, VALUE offset)
+{
+    VALUE markable = *(data_struct + (offset / sizeof(VALUE)));
+    rb_gc_mark_movable(markable);
+}
+
+static inline void
+gc_ref_update_from_offset(rb_objspace_t *objspace, VALUE *data_struct, VALUE offset)
+{
+    #define EDGE (data_struct + (offset / sizeof(VALUE)))
```
Comment (nobu) [inline gc.c:7227]: Why not using this macro in `gc_mark_from_offset` too?
```suggestion
```

File: gc.c [hunk 2]
```diff
@@ -7208,6 +7208,47 @@ gc_mark_imemo(rb_objspace_t *objspace, VALUE obj)
...
+{
+    return (type->flags & RUBY_TYPED_DECL_MARKING) != 0;
+}
+
+static inline void
+gc_mark_from_offset(rb_objspace_t *objspace, VALUE *data_struct, VALUE offset)
+{
+    VALUE markable = *(data_struct + (offset / sizeof(VALUE)));
+    rb_gc_mark_movable(markable);
+}
```
Comment (nobu) [inline gc.c:7222]: - Use `size_t` for offsets.
- Use `EDGE` macro also in this function.

```suggestion
#define EDGE *(VALUE *)((char *)data_struct + offset)

static inline void
gc_mark_from_offset(rb_objspace_t *objspace, void *data_struct, size_t offset)
{
    rb_gc_mark_movable(EDGE);
}
```

File: gc.c [hunk 3]
```diff
@@ -7208,6 +7208,47 @@ gc_mark_imemo(rb_objspace_t *objspace, VALUE obj)
...
+
+static inline void
+gc_ref_update_from_offset(rb_objspace_t *objspace, VALUE *data_struct, VALUE offset)
```
Comment (nobu) [inline gc.c:7225]: - Use `size_t` for offsets

```suggestion
gc_ref_update_from_offset(rb_objspace_t *objspace, void *data_struct, size_t offset)
```

File: gc.c [hunk 4]
```diff
@@ -7313,13 +7354,17 @@ gc_mark_children(rb_objspace_t *objspace, VALUE obj)
         {
             void *const ptr = DATA_PTR(obj);
             if (ptr) {
-                RUBY_DATA_FUNC mark_func = RTYPEDDATA_P(obj) ?
-                    any->as.typeddata.type->function.dmark :
-                    any->as.data.dmark;
-                if (mark_func) (*mark_func)(ptr);
+                if (RTYPEDDATA_P(obj) && gc_declarative_marking_p(any->as.typeddata.type)) {
+                    gc_process_edges(objspace, obj, gc_mark_from_offset);
+                } else {
```
Comment (nobu) [inline gc.c:7359]:
```suggestion
                }
                else {
```

File: gc.c [hunk 5]
```diff
@@ -10633,7 +10678,9 @@ gc_update_object_references(rb_objspace_t *objspace, VALUE obj)
         {
             void *const ptr = DATA_PTR(obj);
             if (ptr) {
-                if (RTYPEDDATA_P(obj)) {
+                if (RTYPEDDATA_P(obj) && gc_declarative_marking_p(any->as.typeddata.type)) {
+                    gc_process_edges(objspace, obj, gc_ref_update_from_offset);
+                } else if (RTYPEDDATA_P(obj)) {
```
Comment (nobu) [inline gc.c:10683]:
```suggestion
                }
                else if (RTYPEDDATA_P(obj)) {
```

File: include/ruby/internal/gc.h [hunk 6]
```diff
@@ -44,6 +44,12 @@
 
 RBIMPL_SYMBOL_EXPORT_BEGIN()
 
+#define REF_EDGE(s, p) (offsetof(struct s, p))
+#define REFS_LIST_PTR(l) ((RUBY_DATA_FUNC)l)
```
Comment (nobu) [inline include/ruby/internal/gc.h:48]: Why these are not `RUBY_` prefixed?

**Comment by nobu** [comment]
https://github.com/ruby/ruby/pull/9078

File: include/ruby/internal/gc.h [hunk 7]
```diff
@@ -44,6 +44,12 @@
...
+#define RUBY_REF_END SIZE_MAX
+#define RUBY_REFERENCES_START(t) static size_t t[] = {
+#define RUBY_REFERENCES_END RUBY_REF_END, };
```
[the remark this answers is not in the document]
Comment (vicspina) [reply include/ruby/internal/gc.h:51]: Hi

**Pull request merged**